lib/url-pattern/url-match-pattern.js

```diff
@@ -0,0 +1,278 @@
+ *
+ * @example <caption>An example UrlMatchPattern</caption>
+ * var matchPattern = new UrlMatchPattern('https://*.google.com/*');
+*/
+_.inherit((
+  UrlMatchPattern = function UrlMatchPattern (pattern) {
+      UrlMatchPattern.super_.apply(this, { pattern: pattern });
+
+      this.pattern = pattern && _.isString(pattern) ? pattern : MATCH_ALL_URLS;
+  }), Property);
+
+_.assign(UrlMatchPattern.prototype, /** @lends UrlMatchPattern.prototype */ {
+    /**
+     * Used to generate the match regex object from the match string we have.
+     * @private
+     * @returns Match regex object
+     */
+    createMatchPattern: function() {
+        var matchPattern = this.pattern,
+
+            // Check the match pattern of sanity and split it into protocol, host and path
+            match = matchPattern.match(regexes.patternSplit);
+
+        if (!match) {
+        // This ensures it is a invalid match pattern
+            return undefined;
+        }
+
+        return ({
+            protocol: match[1],
+            host: match[2],
+            path: this.globPatternToRegexp(match[3])
+        });
+    },
+
+    /**
+     * Converts a given glob pattern into a regular expression
+     * @private
+     * @return {RegExp=}
+     */
+    globPatternToRegexp: function(pattern) {
+        // Escape everything except ? and *.
+        pattern = pattern.replace(regexes.escapeMatcher, regexes.escapeMatchReplacement);
+        pattern = pattern.replace(regexes.questionmarkMatcher, regexes.questionmarkReplacment);
+        pattern = pattern.replace(regexes.starMatcher, regexes.starReplacement);
+        return new RegExp(PREFIX_DELIMITER + pattern + POSTFIX_DELIMITER);
+    },
+
+    /**
+     * Tests if the given protocol string, is allowed by the pattern.
+     *
+     * @param {String=} The protocol to be checked if the pattern allows.
+     * @return {Boolean=}
+     */
+    testProtocol: function(protocol) {
+        var matchRegexObject = this.createMatchPattern();
+        return this._testProtocol(matchRegexObject, protocol);
+    },
+
+    /**
+     * Check for the protocol match for the given url
+     * @private
+     * @param {Object=} [matchRegexObject] The regex object generated by the createMatchPattern function.
+     * @param {String} [protocol] The protocol for which the match needs to be checked.
+     * @return {Boolean}
+     */
+    _testProtocol: function(matchRegexObject, protocol) {
+        // Do not match if protocol is invalid
+        if (!_.includes(ALLOWED_PROTOCOLS, protocol)) {
+            return false;
+        }
+
+        // Return true if test pattern is *, else return match
+        return matchRegexObject.protocol === MATCH_ALL ? true : matchRegexObject.protocol === protocol;
+    },
+
+    /**
+     * Tests if the given host string, is allowed by the pattern.
+     *
+     * @param {String=} The host to be checked if the pattern allows.
+     * @return {Boolean=}
+     */
+    testHost: function(host) {
```

Comment:
The reason for taking in `matchRegexObject` as a parameter, is to avoid the regex parsing multiple times within the same method. So that we could use the same `matchRegexObject` for `_testHost`, `_testProtocol` and `_testPath`.

`testHost` is a nice public API, which creates `matchRegexObject` and wraps `_testHost` for anyone else wanting to use it.